hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/CloseRegionHandler.java

```diff
@@ -38,9 +36,13 @@
 /**
  * Handles closing of a region on a region server.
  * <p/>
- * Now for regular close region request, we will use {@link UnassignRegionHandler} instead. But when
- * shutting down the region server, will also close regions and the related methods still use this
- * class so we keep it here.
+ * In normal operation, we use {@link UnassignRegionHandler} closing Regions but when shutting down
+ * the region server and closing out Regions, we use this handler instead; it does not expect to
```

Comment:
It's really a shame if the only difference between `UnassignRegionHandler` and `CloseRegionHandler` is communication back to the master. should consolidate them if we can.

Reply:
I tried. There are subtle differences ... one takes encoded region name only coming in from rpc. The other can assume it called by the hosting Region.  I played around.  A rewrite would disturb a bunch. I opened "HBASE-24015 Coverage for Assign and Unassign of Regions on RegionServer on failure". I think digging in on this issue will reveal issues around these handlers and will bring on refactor only we'll have extra coverage first so can be confident making change. At the moment afraid to change too much.

Reply:
The work is done by me before. What I could say is that, we have done a lot of hacks only for testing, in CloseRegionHandler, just like we want to do in this patch...

That's why I introduced a new class called UnassignRegionHandler for implementing the region assignment, and still kept the class here. The first idea was to just remove this class but I finally gave up

Reply:
Oof. That's rough @Apache9 .

Reply:
CRH is not for tests only. If you want to remove it, more work needs to be done first. Meantime, lets land this patch to fix a bug around cluster shutdown revealed by tests.